src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs

```diff
@@ -0,0 +1,169 @@
+using System.Diagnostics;
+using System.Collections.Generic;
+
+using Internal.Text;
+using Internal.TypeSystem;
+using Internal.NativeFormat;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Hashtable of all exact (non-canonical) generic method instantiations compiled in the module.
+    /// </summary>
+    internal sealed class ExactMethodInstantiationsNode : ObjectNode, ISymbolNode
+    {
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        private ExternalReferencesTableNode _externalReferences;
+
+        private NativeWriter _nativeWriter;
+        private Section _nativeSection;
+        private VertexHashtable _hashtable;
+
+        private HashSet<MethodDesc> _exactMethodInstantiationsList;
+
+        public ExactMethodInstantiationsNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__exact_method_instantiations_End", true);
+            _externalReferences = externalReferences;
+
+            _nativeWriter = new NativeWriter();
+            _hashtable = new VertexHashtable();
+            _nativeSection = _nativeWriter.NewSection();
+            _nativeSection.Place(_hashtable);
+
+            _exactMethodInstantiationsList = new HashSet<MethodDesc>();
+        }
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__exact_method_instantiations");
+        }
+
+        public ISymbolNode EndSymbol => _endSymbol;
+        public int Offset => 0;
+        public override bool IsShareable => false;
+        public override ObjectNodeSection Section => ObjectNodeSection.DataSection;
+        public override bool StaticDependenciesAreComputed => true;
+        protected override string GetName() => this.GetMangledName();
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node does not trigger generation of other nodes.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+
+            // Zero out the hashset so that we AV if someone tries to insert after we're done.
+            _exactMethodInstantiationsList = null;
+
+            MemoryStream stream = new MemoryStream();
+            _nativeWriter.Save(stream);
+
+            byte[] streamBytes = stream.ToArray();
+
+            _endSymbol.SetSymbolOffset(streamBytes.Length);
+
+            return new ObjectData(streamBytes, Array.Empty<Relocation>(), 1, new ISymbolNode[] { this, _endSymbol });
+        }
+
+        public DependencyList GetExactMethodInstantiationDependenciesForMethod(NodeFactory factory, MethodDesc method)
+        {
+            if (!IsMethodEligibleForTracking(method))
+                return null;
+
+            DependencyList dependencies = new DependencyList();
```

Comment:
I have refactored the APIs on NativeLayoutInfoNode a bit to return a Vertex wrapper node, which provides dependencies information, that various parts of the system can use and know all dependencies associated with the vertex. I hooked it up to this API.
I also added some static APIs on NativeLAyoutInfoNode that compute the dependencies of a certain vertex, without actually producing the vertex or modifying any of the contents of the native layout writer.